### PR TITLE
Adding forfeit method and some tests

### PR DESCRIPTION
**Pull request by rodpat**

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
   root 'static_pages#index'
   resources :games, only: %i[new create show index update] do
+    post 'forfeit', on: :member
```
Comment (mscoutermarsh) [inline config/routes.rb]: I'd go with a `put`. Because it is making a `change`, rather than creating something new.

File: app/models/game.rb [hunk 2]
```diff
@@ -51,4 +54,10 @@ def populate_board
   def self.available
     where('white_player_id IS NULL OR black_player_id IS NULL')
   end
+
+  def forfeit(current_user)
+    update_attributes(winner_id: white_player_id, result: 'Forfeit', black_player_id: nil) if current_user.id == black_player_id
+
+    update_attributes(winner_id: black_player_id, result: 'Forfeit', white_player_id: nil) if current_user.id == white_player_id
+  end
```
Comment (mscoutermarsh) [inline app/models/game.rb]: This looks good.

One change: remove `black_player_id: nil` and `white_player_id: nil`.

A player can be associated as both a black/white player and the winner at the same time.

Comment (mscoutermarsh) [reply app/models/game.rb]: `current_user.id == white_player_id`

Could be `current_user == white_player`

File: app/models/game.rb [hunk 3]
```diff
@@ -2,6 +2,9 @@ class Game < ApplicationRecord
   validates :name, presence: true
   belongs_to :user
   has_many :pieces
+  belongs_to :white_player, class_name: 'User', optional: true
+  belongs_to :black_player, class_name: 'User', optional: true
```
Comment (mscoutermarsh) [inline app/models/game.rb:6]: These associations weren't there before?

File: spec/controllers/games_controller_spec.rb [hunk 4]
```diff
@@ -74,4 +74,28 @@
...
   end
+
+  describe 'games#forfeit action' do
+    it 'should show white player as winner if black player forfeits' do
+      white = FactoryBot.create(:user)
+      black = FactoryBot.create(:user)
+      sign_in white
+      game = FactoryBot.create(:game)
+      game.update_attributes(white_player_id: white.id, black_player_id: black.id)
+      post :forfeit, params: { id: game.id, game: { winner_id: white.id } }
```
Comment (mscoutermarsh) [inline spec/controllers/games_controller_spec.rb]: ` put :forfeit, params: { id: game.id }`

File: app/controllers/games_controller.rb [hunk 5]
```diff
@@ -36,6 +36,13 @@ def update
     end
   end
 
+  def forfeit
+    @game = Game.find_by(id: params[:id])
+    @game.forfeit(current_user)
+    flash[:alert] = 'You have forfeited the game.'
+    redirect_to games_path
```
Comment (dynnog) [inline app/controllers/games_controller.rb:43]: I'm thinking we can add a modal in the game_show_page to create a pop-up message that verify the user if they want to forfeit. If they choose to forfeit again, then we do `redirect_to games_path(current_game), notice: "You have forfeited the game"`.

File: config/routes.rb [hunk 6]
```diff
@@ -3,6 +3,7 @@
 
   root 'static_pages#index'
   resources :games, only: %i[new create show index update] do
+    put 'forfeit', on: :member
```
Comment (dynnog) [inline config/routes.rb:6]: Is this line supposed to be `member`? or `user`?

File: spec/controllers/games_controller_spec.rb [hunk 7]
```diff
@@ -74,4 +74,28 @@
       expect(response).to have_http_status(:not_found)
     end
   end
+
+  describe 'games#forfeit action' do
+    it 'should show white player as winner if black player forfeits' do
+      white = FactoryBot.create(:user)
+      black = FactoryBot.create(:user)
+      sign_in white
```
Comment (dynnog) [inline spec/controllers/games_controller_spec.rb:82]: Did you mean `sign_in black`?

File: app/views/games/show.html.erb [hunk 8]
```diff
@@ -18,4 +18,5 @@
   <% end %>
 </table>
 <br />
+<%= link_to 'Forfiet', forfeit_game_path, class: 'btn btn-danger', method: :PUT %>
```
Comment (dynnog) [inline app/views/games/show.html.erb]: We can add a modal here that links to the Forfeit. We can also alter it to verify `if current_user != game.user` then `alert: 'You are not part of the game!`.

File: config/routes.rb (hunk 6 above)
Comment (gamalhetari) [reply config/routes.rb:6]: I asked my mentor about it he said member was fine.

File: spec/controllers/games_controller_spec.rb (hunk 7 above)
Comment (gamalhetari) [reply spec/controllers/games_controller_spec.rb:82]: No sign in white because whoever creates the game becomes the white player.

**Pull request merged**